src/Enzo/enzo_EnzoBfieldMethodCT.cpp

```diff
@@ -261,10 +260,10 @@ void EnzoBfieldMethodCT::compute_center_efield
   int k = coord.k_axis();
 
   // Load the jth and kth components of the velocity and cell-centered bfield
-  EFlt3DArray velocity_j = prim_map.at(v_names[j]);
-  EFlt3DArray velocity_k = prim_map.at(v_names[k]);
-  EFlt3DArray bfield_j = prim_map.at(b_names[j]);
-  EFlt3DArray bfield_k = prim_map.at(b_names[k]);
+  EFlt3DArray velocity_j = integration_map.at(v_names[j]);
+  EFlt3DArray velocity_k = integration_map.at(v_names[k]);
+  EFlt3DArray bfield_j = integration_map.at(b_names[j]);
+  EFlt3DArray bfield_k = integration_map.at(b_names[k]);
```

Comment:
I've thought about this exact sort of change a lot, and I'm not sure whether this is an improvement or not.

The thing is that EFlt3DArray can be thought of as a shared pointer. By making a copy of it (what we have now), we pay the overhead of incrementing/decrementing the reference count (although the current case probably uses move semantics). By storing a reference to it, we avoid increments/decrements, but now we effectively have a reference to a pointer (which is similar to a pointer to a pointer). I'm not sure what's better

An alternative idea is just to mark ``velocity_j``, ``velocity_k``, ``bfield_j``, and ``bfield_k`` as `const`.

The same sort of semantics apply to ``Kokkos::Views``, so I'm willing to go with your instinct here. But before I change it, do you have anything else to add?.

Reply:
> The thing is that EFlt3DArray can be thought of as a shared pointer. By making a copy of it (what we have now), we pay the overhead of incrementing/decrementing the reference count (although the current case probably uses move semantics). By storing a reference to it, we avoid increments/decrements, but now we effectively have a reference to a pointer (which is similar to a pointer to a pointer). I'm not sure what's better

We've been using essentially `const Kokkos:View&` recently, I appreciate that we keep the same object. Sometimes it can make debugging easier since the address of the container doesn't change between contexts.

The other confusing piece to me is that all of our arrays can probably be made `const &`, since we even though we're changing data in the array we're not changing the data itself.

> The same sort of semantics apply to `Kokkos::Views`, so I'm willing to go with your instinct here. But before I change it, do you have anything else to add?.

I'm not planning on adding any more suggestions. I would suggest to try batching all of my suggestions together into one commit via `github`, then see if it compiles and runs. I've probably missed a few things that would also need to be constant but that should be straightfoward to find via the compiler. Worst case you could undo the commit entirely.

Reply:
> The other confusing piece to me is that all of our arrays can probably be made const &, since we even though we're changing data in the array we're not changing the data itself

That's part of the reason why I've largely avoided referring to `EFlt3DArray` as a `const` in most places yet.

When I originally developed `CelloArray`, a lot of it's design was strongly motivated by my familiarity with NumPy. I wasn't fond of the way that traditional "container-semantics" will produce a deepcopy when you perform an assignment (it also makes subarray creation more confusing). It was originally my intention to prevent modification of the contents for `const CelloArray`, but that led to some strange quirks in the Copy-Constructor and Copy-Assignment operation. (Some stuff with temporaries break. It's confusing to new users why certain things break, but I don't think it actually limits what we want to do in the main codebase)

At around the time that I was struggling with this, I started reading about `Kokkos` and I realized that `CelloArray` has pointer-semantics just like `Kokkos::View` (a better name for `CelloArray` is probably `CelloView`), so I decided to lean into it.

I'm working on something that might help the situation. I think I mentioned this before, but I've slowly been working on a new PR, in which I introduce support for `CelloArray<const T, D>` which prevents the contents of the array from being mutated (I needed to make sure that this case element access works right support and add code to facilitate casts from `CelloArray<T, D>` and `CelloArray<const T, D>`). I got a little side-tracked over the past 2 weeks, but I hope to get back to it soon-ish. Once I'm done, it's my intention to alias  `CelloArray<const enzo_float, D>` as something like `RdOnlyEFlt3DArray` (I'm definitely open to better names) and start introducing that throughout the VL+CT integrator.

But, It might be worthwhile to discuss changing the `CelloArray` semantics: 
- If you and @pgrete, don't think it would be confusing, it might be worth revisiting the idea of making `const CelloArray` totally immutable. I'm not convinced that it's completely impossible. It might be easier now that I have a working test suite and I'm planning to remove the bulk assignment stuff from the `CelloArray` API (which will drastically simplify the implementation).
- I'd also be willing to consider a shift to container semantics (I don't think this would break anything). But if we did that, I would want some help with improving

Reply:
Just as an aside, the aforementioned support for `CelloArray<const T, D>` is being implemented as part of PR #113

Reply:
For the `EFlt3DArray` arrays that are currently used, I do think using `const  EFlt3DArray&` is preferable to `EFlt3DArray` since the array shouldn't be changed even if the data is changed, and using the compiler to enforce that is preferable.

> I'm working on something that might help the situation. I think I mentioned this before, but I've slowly been working on a new PR, in which I introduce support for `CelloArray<const T, D>` which prevents the contents of the array from being mutated (I needed to make sure that this case element access works right support and add code to facilitate casts from `CelloArray<T, D>` and `CelloArray<const T, D>`).

This would be the preferable syntax so that it's clear the array points to `const` data.

> I got a little side-tracked over the past 2 weeks, but I hope to get back to it soon-ish. Once I'm done, it's my intention to alias `CelloArray<const enzo_float, D>` as something like `RdOnlyEFlt3DArray` (I'm definitely open to better names) and start introducing that throughout the VL+CT integrator.

I'd lean against aliasing the type. The code would be more explicit using `CelloArray<const enzo_float, D>` and `CelloArray<enzo_float, D>` everywhere so that it's clear that it's a `CelloArray` and references const or non-const data.

>     * If you and @pgrete, don't think it would be confusing, it might be worth revisiting the idea of making `const CelloArray` totally immutable. I'm not convinced that it's completely impossible. It might be easier now that I have a working test suite and I'm planning to remove the bulk assignment stuff from the `CelloArray` API (which will drastically simplify the implementation).

Make it so `const CelloArray` means the data it points to can't be modified? I'm against that approach, I think it makes sense to preserve `CelloArray<const T, D>` vs. `const CelloArray<T, D>` type.